src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3178,6 +3178,10 @@ private void VisitLocalFunctionUse(LocalFunctionSymbol symbol)
                 // If the starting state of the local function has changed and we've already visited
                 // the local function, we need another pass
                 stateChangedAfterUse = true;
+
+                // Local function starting state has changed, so a loop inside the local function
+                // can have an outdated head state, hence we need to recompute it in the next pass.
```

Comment:
For example, why should the loops outside of the local function be affected?

It looks like ```AnalyzeLocalFunctionOrLambda``` helper saves the current state of the visitor, prepares the "local function state", visits the nested body, and then restores the original state. Perhaps the loop head state should also be a part of the save/initialize/restore process. 
I also noticed another strange thing about what ```AnalyzeLocalFunctionOrLambda``` is doing. It appears that it doesn't preserve any new "learnings" from visiting the body. This is probably the reason why no warnings reported for ([https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwBYBuc44gYhlQBs2Jg2BTcnmF16MSAZko5yAYU4BnWYSIBvYgEgqANkrVyAWWgwAFAEo1KoqtUA3CMnKyeCcgF5yMHgHdyACQiyAFgDKjgA8AOKoUAAmAHwmtCKWYI7+APZRePGJ6jrJCGlROCZqquaWqg5OrqwcCRaqAL7ZWLkp6Zmm9WVJbYVZROSDQ8ODlQB0Mjy2/ZZNFnMNQA](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsAGAAizwBYBuc44gYhlQBs2Jg2BTcnmF16MSAZko5yAYU4BnWYSIBvYgEgqANkrVyAWWgwAFAEo1KoqtUA3CMnKyeCcgF5yMHgHdyACQiyAFgDKjgA8AOKoUAAmAHwmtCKWYI7+APZRePGJ6jrJCGlROCZqquaWqg5OrqwcCRaqAL7ZWLkp6Zmm9WVJbYVZROSDQ8ODlQB0Mjy2/ZZNFnMNQA), but the code throws ```System.NullReferenceException```:
```
using System;
using System.Collections.Generic; 

#nullable enable

public class Class1
{
 static void Main()
 {
  var set = new HashSet<Guid>();

  method1();

  void method2()
  {
   set = null;
  }

  void method1()
  {
   method2();
            set.Clear();
  }
 }
}
```

Reply:
> I also noticed another strange thing about what `AnalyzeLocalFunctionOrLambda` is doing. It appears that it doesn't preserve any new "learnings" from visiting the body.

I believe it is by design that nullable state flows in to local functions but does not flow out. However, I can't find any documentation offhand about this. IIRC, it was thought that letting state both flow in and out would cause us to need too many analysis passes to reach a stable state.

Reply:
> Why loop head state is so special? Do we clear any other state that possibly got accumulated?

Seems special to me, because it relies on running multiple passes, i.e., the loop head state is reused from previous passes to simulate the loop. I'm not sure if there are other states like that, but if so, then yes, they should be cleaned up in the same way.

> Perhaps the loop head state should also be a part of the save/initialize/restore process.

I don't see how that would help. The loop head state is incorrect here (in the example in the issue), as the local function is visited first as definition (since there's no use of it in the method body) with "unreachable" state and hence it sees the variable as nullable, which will be recorded in the loop head state and that will be reused in all following passes. That's why I clear the loop head state after the local function *use* is visited (where the starting state of the local function is changed to see the variable as non-null, correctly).

Another possibility I considered is fixing the fact that even though the local function definition is visited with unreachable state, it could perhaps see the variable as non-null - but I imagine there are good reasons it is not implemented like that.

> why should the loops outside of the local function be affected?

Doesn't have to be, but since we will be running more passes anyway, it seems like it doesn't matter.

I could try saving the loop head state inside the local function's state and clear just that if you think that's better. I think I would need to
1. Create a `localFunctionState.LoopHeadState` field.
2. Set `AbstractFlowPass._loopHeadState` to `localFunctionState.LoopHeadState` before visiting its body and resetting it back afterwards.
3. Clear `localFunctionState.LoopHeadState` at this place (where I currently set `clearLoopHeadState = true;`).

Reply:
> By this logic, any state can/should be thrown away since we will be running more passes anyway. However, we are not doing that. There is a good reason why some state is transferred from one pass to another. In general, an extra pass doesn't completely invalidate result of previous passes, it is additive.

True, I meant more that if we are deleting one loop state, we can delete them all since they behave analogously and will need the same number of passes to recompute. But yes, I'm not sure that's correct, so I will try to limit this change. Thanks. We obviously cannot delete any state, that would lead to infinite loop - we have to at least preserve the local function starting state that has just changed and is the reason we want to re-run.

> Before trying to fix a problem, we should have a good understanding of the problem. Under what conditions it occurs, what specific portion of data causing a problem, when and where the data populated, what should be the lifetime of data and why, etc. The PR description provides no details about that.

I wrote a comment in the code, hoping that's better than PR description as it's useful also for future readers of the code. Of course, that might not be enough to understand the problem. I have tried to describe it also in my previous comment, but I can provide more details. To sum up, this is happening in the example in the issue:
1. `method2` is visited as definition. The local function starting state is unreachable at this point. `s` is nullable. That's also stored in the head loop state.
2. `method1` is visited and inside it, `method2` use is visited. There, the local function starting state changes to reachable and `s` is non-null. However, when we visit the loop, its previous loop head state is reused (to "emulate the looping"), marking `s` as nullable.

Reply:
I guess we are still lacking an understanding or an explanation of: " what should be the lifetime of data and why, etc." Basically we are lacking a general explanation that might look like: analysis is supposed/designed to work this way ..., data originate and flow this way ..., at a certain point they become obsolete (because such and such), keeping them around beyond this point is causing problems like ..., dropping them at this point in time isn't going to cause new problems because ...     
This information should target a more generalized case, not just a simplistic repro scenario. The feeling am looking for after getting all the information:
- we understand how the machinery works
- we understand where and why things go wrong
- we understand what should be done instead and why that is not going to have a negative effect elsewhere
- the proposed change precisely corresponds to the previous item

Reply:
We chatted about this in stand up and our take away is that we need to sync up and get a better understanding of how this code is supposed to work: both from a spec level understanding as well as implementation. There were a few people missing from standup today but no one there felt comfortable asserting ownership of this area enough to guide the PR. I will grab some time next week for us to chat about this.